Refactored plugin for performance and modern coding standards.

sun opened the pull request:
Internal ticket: https://app.asana.com/0/30975952354191/247630419160473/f

Goal
* Detach the Varnish integration from the varnish-http-purge mainline to enable more improvements and prevent false updates, as it has significantly diverged from it.

Procedure
1. Refactor the code.
1. Document what needs to changed/upgraded on existing sites in changelog (readme.txt).

sun committed:
Refactored plugin for performance and modern coding standards.

sun commented on src/Varnish.php:
```diff
@@ -0,0 +1,77 @@
+    $schema = defined('VARNISH_SCHEMA') ? VARNISH_SCHEMA : 'http://';
+
+    if (defined('VARNISH_IP')) {
+      $varnish_purge_url = $schema . VARNISH_IP . $path;
```
A typical definition might include a port, which is optional, but possible, and not uncommon (whenever Varnish is hosted on the same machine as the webserver):
```php
const VARNISH_IP = '127.0.0.1:8080';
```
Not particularly happy with introducing this as an "IP" instead of e.g. "HOST".  Opinions?

Suggested change:
```php
const VARNISH_HOST = '127.0.0.1:8080';
```
and because the port is tightly coupled to the protocol:
```php
const VARNISH_HOST = 'http://127.0.0.1:8080';
```

_Background: Maxcluster supports to run Varnish on port 8000 (SSL) or 8080 (non-SSL), but even if it runs with SSL, the internal port for PURGE requests is 8080 (and thus also HTTP)._

This would additionally get rid of the separate `VARNISH_SCHEMA` constant which is optionally used just a few lines above.

fabianmarz replied:
IMO using the `VARNISH_HOST` constant is a better approach as we can get rid of  another constant. As the port is coupled to the protocol anyway there should not be any downside, correct?

fabianmarz commented on readme.txt:
```diff
@@ -1,233 +1,199 @@
-== Installation ==
-No WordPress configuration needed.
+2. Configure the IP and optionally the port of your Varnish server in the
+   `VARNISH_IP` constant in your `wp-config.php`:
```
Make sure to adjust this constant (and further in the file) as well when switching to `VARNISH_HOST`.

fabianmarz commented on src/Admin.php:
```diff
@@ -0,0 +1,109 @@
+      add_action('admin_bar_menu', __CLASS__ . '::admin_bar_menu'), 100);
+    }
+    // Output manual purge button on admin dashboard.
+    add_action('activity_box_end', __CLASS__ . '::activity_box_end', 100);
```
Is ensured this is visible to admins with the correct privileges only? Probably it makes sense to move it into the `hasPurgeAccess` check as well?

sun replied:
The meta box outputs a message about missing access privileges currently. We probably want to revisit the UI/UX in the future, but I don't think we should do that as part of this PR.

fabianmarz commented on src/Plugin.php:
```diff
@@ -0,0 +1,194 @@
+
+    add_action('switch_theme', __CLASS__ . '::purgePost', 10, 2);
+
+    add_action('wp_update_attachment_metadata', __CLASS__ . '::purgeAttachmentMeta', 50, 2);
```
Why are you using a custom `camelCase` class method here, while using a `snake_case` method calls [here](https://github.com/netzstrategen/wordpress-varnish/pull/1/files#diff-4fb119e14a4976b590190226e5b0075aR21)?

sun replied:
Because multiple hooks are registering the same callback, so there is not a single hook name to use as base for a method name.  The snake_case is used when there is a 1:1 relationship between hook and callback.

fabianmarz replied:
`purgeAttachmentMeta` is used by `wp_update_attachment_metadata` only. Anyway, I think we can proceed with the current solution. In case another hook makes use of this as well, we don't have to adjust this in multiple places.

sun replied:
Ah, I see what you mean now.

fabianmarz commented on src/Plugin.php:
```diff
@@ -0,0 +1,194 @@
+   */
+  public static function purgePost($postId) {
+    // Skip post revisions.
+    if (get_permalink($postId) === FALSE) {
```
IMO it makes sense to store the permalink in a variable to avoid multiple function calls.

fabianmarz commented on src/Plugin.php:
```diff
@@ -0,0 +1,194 @@
+
+    // Author URL.
+    $author_id = get_post_field('post_author', $postId);
+    Varnish::$purgeUrls[$author_url = get_author_posts_url($author_id)] = FALSE;
```
While this is a handy writing, I'd suggest to move the `$author_url` to a separate line to keep consistency and maintain easy readability (similar to https://github.com/netzstrategen/wordpress-varnish/pull/1/files#diff-f7fcf01b75ab0236c970de527853c3abR76)

fabianmarz commented on src/Varnish.php:
```diff
@@ -0,0 +1,77 @@
+class Varnish {
+
+  /**
+   * URLs to purge in Varnish whose values are Booleans indicating regex.
```
Probably state it's an associative array whose keys are the URLs and the values are the Booleans indicating regex.

fabianmarz commented on src/Varnish.php:
```diff
@@ -0,0 +1,77 @@
+   *   Whether $url is a regular expression.
+   */
+  public static function purgeUrl($url, $is_regex = FALSE) {
+    $headers = array();
```
Why not using the array shorthand syntax here as some lines above?

sun commented on src/Plugin.php:
```diff
@@ -0,0 +1,196 @@
+    add_action('edit_post', __CLASS__ . '::purgePost', 10, 2);
+    add_action('delete_attachment', __CLASS__ . '::purgePost', 10, 2);
+
+    add_action('switch_theme', __CLASS__ . '::purgePost', 10, 2);
```
[switch_theme](https://developer.wordpress.org/reference/functions/switch_theme/) does not get a post ID as first argument, so this preexisting code is rather pointless.

sun replied:
Fixed in last commit.

fabianmarz reviewed (approved):
To me looks good now! Really nice work 🎉 

Please note, I haven't tested the functionality, but I assume it works as expected as we're using this on several pages already and you checked during development/refactoring.

sun committed:
Removed manual cache purge/flush menu link in admin bar.

The cache should work silently in the background. Manual cache flushes should be an exception. Site administrators should not have to care for cache invalidation in their daily routines. Therefore, the admin bar link is superfluous noise.
